File: library/src/main/java/com/zhongjh/cameraviewsoundrecorder/settings/GlobalSetting.java
package com.zhongjh.cameraviewsoundrecorder.settings;

import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.annotation.StyleRes;
import android.support.v4.app.Fragment;

import com.zhongjh.cameraviewsoundrecorder.MainActivity;
import com.zhongjh.cameraviewsoundrecorder.album.engine.ImageEngine;
import com.zhongjh.cameraviewsoundrecorder.album.engine.impl.GlideEngine;
import com.zhongjh.cameraviewsoundrecorder.album.engine.impl.PicassoEngine;
import com.zhongjh.cameraviewsoundrecorder.album.enums.MimeType;
import com.zhongjh.cameraviewsoundrecorder.album.filter.Filter;
import com.zhongjh.cameraviewsoundrecorder.album.listener.OnCheckedListener;
import com.zhongjh.cameraviewsoundrecorder.album.listener.OnSelectedListener;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Set;

import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_BEHIND;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_FULL_USER;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_LOCKED;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_NOSENSOR;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_PORTRAIT;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_SENSOR;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_SENSOR_PORTRAIT;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_USER;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_USER_LANDSCAPE;
import static android.content.pm.ActivityInfo.SCREEN_ORIENTATION_USER_PORTRAIT;

/**
 * 用于构建媒体具体设置 API。
 * Created by zhongjh on 2018/9/28.
 */
public final class GlobalSetting {

    private final MultiMedia mMultiMedia;
    private final GlobalSpec mGlobalSpec;

    // www.代替枚举的@IntDef用法
    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
    @IntDef({
            SCREEN_ORIENTATION_UNSPECIFIED,
            SCREEN_ORIENTATION_LANDSCAPE,
            SCREEN_ORIENTATION_PORTRAIT,
            SCREEN_ORIENTATION_USER,
            SCREEN_ORIENTATION_BEHIND,
            SCREEN_ORIENTATION_SENSOR,
            SCREEN_ORIENTATION_NOSENSOR,
            SCREEN_ORIENTATION_SENSOR_LANDSCAPE,
            SCREEN_ORIENTATION_SENSOR_PORTRAIT,
            SCREEN_ORIENTATION_REVERSE_LANDSCAPE,
            SCREEN_ORIENTATION_REVERSE_PORTRAIT,
            SCREEN_ORIENTATION_FULL_SENSOR,
            SCREEN_ORIENTATION_USER_LANDSCAPE,
            SCREEN_ORIENTATION_USER_PORTRAIT,
            SCREEN_ORIENTATION_FULL_USER,
            SCREEN_ORIENTATION_LOCKED
    })
    @Retention(RetentionPolicy.SOURCE)
    @interface ScreenOrientation {
    }

    /**
     * Constructs a new specification builder on the context.
     *
     * @param multiMedia   a requester context wrapper.
     * @param mimeTypes MIME type set to select.
     */
    GlobalSetting(MultiMedia multiMedia, @NonNull Set<MimeType> mimeTypes, boolean mediaTypeExclusive) {
        mMultiMedia = multiMedia;
        mGlobalSpec = GlobalSpec.getCleanInstance();
        mGlobalSpec.mimeTypeSet = mimeTypes;
        mGlobalSpec.mediaTypeExclusive = mediaTypeExclusive;
        mGlobalSpec.orientation = SCREEN_ORIENTATION_UNSPECIFIED;
    }

    public GlobalSetting cameraSetting(CameraSetting cameraSetting){
        mGlobalSpec.cameraSetting = cameraSetting;
        return this;
    }

    /**
     * Whether to show only one media type if choosing medias are only images or videos.
     *
     * @param showSingleMediaType whether to show only one media type, either images or videos.
     * @return {@link GlobalSetting} for fluent API.
     * @see GlobalSpec#onlyShowImages()
     * @see GlobalSpec#onlyShowVideos()
     */
    public GlobalSetting showSingleMediaType(boolean showSingleMediaType) {
        mGlobalSpec.showSingleMediaType = showSingleMediaType;
        return this;
    }

    /**
     * Theme for media selecting Activity.
     * <p>
     * There are two built-in themes:
     * 1. com.zhihu.matisse.R.style.Matisse_Zhihu;
     * 2. com.zhihu.matisse.R.style.Matisse_Dracula
     * you can define a custom theme derived from the above ones or other themes.
     *
     * @param themeId theme resource id. Default value is com.zhihu.matisse.R.style.Matisse_Zhihu.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting theme(@StyleRes int themeId) {
        mGlobalSpec.themeId = themeId;
        return this;
    }

    /**
     * Show a auto-increased number or a check mark when user select media.
     *
     * @param countable true for a auto-increased number from 1, false for a check mark. Default
     *                  value is false.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting countable(boolean countable) {
        mGlobalSpec.countable = countable;
        return this;
    }

    /**
     * Maximum selectable count.
     *
     * @param maxSelectable Maximum selectable count. Default value is 1.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting maxSelectable(int maxSelectable) {
        if (maxSelectable < 1)
            throw new IllegalArgumentException("maxSelectable must be greater than or equal to one");
        if (mGlobalSpec.maxImageSelectable > 0 || mGlobalSpec.maxVideoSelectable > 0)
            throw new IllegalStateException("already set maxImageSelectable and maxVideoSelectable");
        mGlobalSpec.maxSelectable = maxSelectable;
        return this;
    }

    /**
     * Only useful when {@link GlobalSpec#mediaTypeExclusive} set true and you want to set different maximum
     * selectable files for image and video media types.
     *
     * @param maxImageSelectable Maximum selectable count for image.
     * @param maxVideoSelectable Maximum selectable count for video.
     * @return
     */
    public GlobalSetting maxSelectablePerMediaType(int maxImageSelectable, int maxVideoSelectable) {
        if (maxImageSelectable < 1 || maxVideoSelectable < 1)
            throw new IllegalArgumentException(("max selectable must be greater than or equal to one"));
        mGlobalSpec.maxSelectable = -1;
        mGlobalSpec.maxImageSelectable = maxImageSelectable;
        mGlobalSpec.maxVideoSelectable = maxVideoSelectable;
        return this;
    }

    /**
     * Add filter to filter each selecting item.
     *
     * @param filter {@link Filter}
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting addFilter(@NonNull Filter filter) {
        if (mGlobalSpec.filters == null) {
            mGlobalSpec.filters = new ArrayList<>();
        }
        if (filter == null) throw new IllegalArgumentException("filter cannot be null");
        mGlobalSpec.filters.add(filter);
        return this;
    }

    /**
     * Determines whether the photo capturing is enabled or not on the media grid view.
     * <p>
     * If this value is set true, photo capturing entry will appear only on All Media's page.
     *
     * @param enable Whether to enable capturing or not. Default value is false;
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting capture(boolean enable) {
        mGlobalSpec.capture = enable;
        return this;
    }

    /**
     * Show a original photo check options.Let users decide whether use original photo after select
     *
     * @param enable Whether to enable original photo or not
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting originalEnable(boolean enable) {
        mGlobalSpec.originalable = enable;
        return this;
    }

    /**
     * Maximum original size,the unit is MB. Only useful when {link@originalEnable} set true
     *
     * @param size Maximum original size. Default value is Integer.MAX_VALUE
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting maxOriginalSize(int size) {
        mGlobalSpec.originalMaxSize = size;
        return this;
    }

    /**
     * Capture strategy provided for the location to save photos including internal and external
     * storage and also a authority for {@link android.support.v4.content.FileProvider}.
     *
     * @param captureStrategy {@link CaptureStrategy}, needed only when capturing is enabled.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting captureStrategy(CaptureStrategy captureStrategy) {
        mGlobalSpec.captureStrategy = captureStrategy;
        return this;
    }

    /**
     * Set the desired orientation of this activity.
     *
     * @param orientation An orientation constant as used in {@link ScreenOrientation}.
     *                    Default value is {@link android.content.pm.ActivityInfo#SCREEN_ORIENTATION_PORTRAIT}.
     * @return {@link GlobalSetting} for fluent API.
     * @see Activity#setRequestedOrientation(int)
     */
    public GlobalSetting restrictOrientation(@ScreenOrientation int orientation) {
        mGlobalSpec.orientation = orientation;
        return this;
    }

    /**
     * Set a fixed span count for the media grid. Same for different screen orientations.
     * <p>
     * This will be ignored when {@link #gridExpectedSize(int)} is set.
     *
     * @param spanCount Requested span count.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting spanCount(int spanCount) {
        if (spanCount < 1) throw new IllegalArgumentException("spanCount cannot be less than 1");
        mGlobalSpec.spanCount = spanCount;
        return this;
    }

    /**
     * Set expected size for media grid to adapt to different screen sizes. This won't necessarily
     * be applied cause the media grid should fill the view container. The measured media grid's
     * size will be as close to this value as possible.
     *
     * @param size Expected media grid size in pixel.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting gridExpectedSize(int size) {
        mGlobalSpec.gridExpectedSize = size;
        return this;
    }

    /**
     * Photo thumbnail's scale compared to the View's size. It should be a float value in (0.0,
     * 1.0].
     *
     * @param scale Thumbnail's scale in (0.0, 1.0]. Default value is 0.5.
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting thumbnailScale(float scale) {
        if (scale <= 0f || scale > 1f)
            throw new IllegalArgumentException("Thumbnail scale must be between (0.0, 1.0]");
        mGlobalSpec.thumbnailScale = scale;
        return this;
    }

    /**
     * Provide an image engine.
     * <p>
     * There are two built-in image engines:
     * 1. {@link GlideEngine}
     * 2. {@link PicassoEngine}
     * And you can implement your own image engine.
     *
     * @param imageEngine {@link ImageEngine}
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting imageEngine(ImageEngine imageEngine) {
        mGlobalSpec.imageEngine = imageEngine;
        return this;
    }

    /**
     * Set listener for callback immediately when user select or unselect something.
     * <p>
     * It's a redundant API with {@link MultiMedia#obtainResult(Intent)},
     * we only suggest you to use this API when you need to do something immediately.
     *
     * @param listener {@link OnSelectedListener}
     * @return {@link GlobalSetting} for fluent API.
     */
    @NonNull
    public GlobalSetting setOnSelectedListener(@Nullable OnSelectedListener listener) {
        mGlobalSpec.onSelectedListener = listener;
        return this;
    }

    /**
     * Set listener for callback immediately when user check or uncheck original.
     *
     * @param listener {@link OnSelectedListener}
     * @return {@link GlobalSetting} for fluent API.
     */
    public GlobalSetting setOnCheckedListener(@Nullable OnCheckedListener listener) {
        mGlobalSpec.onCheckedListener = listener;
        return this;
    }

    /**
     * Start to select media and wait for result.
     *
     * @param requestCode Identity of the request Activity or Fragment.
     */
    public void forResult(int requestCode) {
        Activity activity = mMultiMedia.getActivity();
        if (activity == null) {
            return;
        }

        Intent intent = new Intent(activity, MainActivity.class);

        Fragment fragment = mMultiMedia.getFragment();
        if (fragment != null) {
            fragment.startActivityForResult(intent, requestCode);
        } else {
            activity.startActivityForResult(intent, requestCode);
        }
    }

}

File: multilibrary/src/main/java/com/zhongjh/multimedia/settings/GlobalSetting.kt
package com.zhongjh.multimedia.settings

import android.app.Activity
import android.content.Intent
import android.content.pm.ActivityInfo
import androidx.activity.result.ActivityResultLauncher
import androidx.annotation.IntDef
import androidx.annotation.StyleRes
import com.zhongjh.multimedia.MainActivity
import com.zhongjh.multimedia.R
import com.zhongjh.common.engine.ImageEngine
import com.zhongjh.multimedia.listener.OnImageCompressionListener
import com.zhongjh.multimedia.listener.OnLogListener
import com.zhongjh.multimedia.preview.PreviewActivity
import com.zhongjh.multimedia.preview.enum.PreviewType
import com.zhongjh.multimedia.preview.start.PreviewSetting
import com.zhongjh.multimedia.settings.GlobalSpec.cleanInstance
import com.zhongjh.multimedia.settings.api.GlobalSettingApi
import com.zhongjh.common.coordinator.VideoCompressCoordinator
import com.zhongjh.common.entity.GridMedia
import com.zhongjh.common.entity.LocalMedia
import com.zhongjh.common.enums.MimeType
import java.lang.ref.WeakReference
import java.util.*

/**
 * 用于构建媒体具体公共设置 API。
 *
 * @author zhongjh
 * @date 2018/9/28
 *
 * 在上下文中构造新的规范生成器。
 *
 * @param multiMediaSetting 在 requester context wrapper.
 * @param mimeTypes         设置为选择的 [MimeType] 类型
 */
class GlobalSetting internal constructor(
    private val multiMediaSetting: MultiMediaSetting,
    mimeTypes: Set<MimeType>
) : GlobalSettingApi {

    private val mGlobalSpec: GlobalSpec = cleanInstance

    @IntDef(value = [ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED, ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE, ActivityInfo.SCREEN_ORIENTATION_PORTRAIT, ActivityInfo.SCREEN_ORIENTATION_USER, ActivityInfo.SCREEN_ORIENTATION_BEHIND, ActivityInfo.SCREEN_ORIENTATION_SENSOR, ActivityInfo.SCREEN_ORIENTATION_NOSENSOR, ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE, ActivityInfo.SCREEN_ORIENTATION_SENSOR_PORTRAIT, ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE, ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT, ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR, ActivityInfo.SCREEN_ORIENTATION_USER_LANDSCAPE, ActivityInfo.SCREEN_ORIENTATION_USER_PORTRAIT, ActivityInfo.SCREEN_ORIENTATION_FULL_USER, ActivityInfo.SCREEN_ORIENTATION_LOCKED])
    @Retention(AnnotationRetention.SOURCE)
    annotation class ScreenOrientation

    override fun onDestroy() {
        mGlobalSpec.activityResultLauncher = null
        if (mGlobalSpec.albumSetting != null) {
            mGlobalSpec.albumSetting!!.onDestroy()
        }
        if (mGlobalSpec.cameraSetting != null) {
            mGlobalSpec.cameraSetting!!.onDestroy()
        }
        mGlobalSpec.onLogListener = null
    }

    override fun albumSetting(albumSetting: AlbumSetting): GlobalSetting {
        mGlobalSpec.albumSetting = albumSetting
        return this
    }

    override fun cameraSetting(cameraSetting: CameraSetting): GlobalSetting {
        mGlobalSpec.cameraSetting = cameraSetting
        return this
    }

    override fun recorderSetting(recorderSetting: RecorderSetting): GlobalSetting {
        mGlobalSpec.recorderSetting = recorderSetting
        return this
    }

    override fun theme(@StyleRes themeId: Int): GlobalSetting {
        mGlobalSpec.themeId = themeId
        return this
    }

    override fun defaultPosition(position: Int): GlobalSetting {
        mGlobalSpec.defaultPosition = position
        return this
    }

    override fun maxSelectablePerMediaType(
        maxSelectable: Int?,
        maxImageSelectable: Int?,
        maxVideoSelectable: Int?,
        maxAudioSelectable: Int?,
        alreadyImageCount: Int,
        alreadyVideoCount: Int,
        alreadyAudioCount: Int
    ): GlobalSetting {
        check(!(maxSelectable == null && maxImageSelectable == null)) { "maxSelectablePerMediaType 方法中如果 maxSelectable 为null，那么 maxImageSelectable 必须是0或者0以上数值" }
        check(!(maxSelectable == null && maxVideoSelectable == null)) { "maxSelectablePerMediaType 方法中如果 maxSelectable 为null，那么 maxVideoSelectable 必须是0或者0以上数值" }
        check(!(maxSelectable == null && maxAudioSelectable == null)) { "maxSelectablePerMediaType 方法中如果 maxSelectable 为null，那么 maxAudioSelectable 必须是0或者0以上数值" }
        check(!(maxSelectable != null && maxImageSelectable != null && maxImageSelectable > maxSelectable)) { "maxSelectable 必须比 maxImageSelectable 大" }
        check(!(maxSelectable != null && maxVideoSelectable != null && maxVideoSelectable > maxSelectable)) { "maxSelectable 必须比 maxVideoSelectable 大" }
        check(!(maxSelectable != null && maxAudioSelectable != null && maxAudioSelectable > maxSelectable)) { "maxSelectable 必须比 maxAudioSelectable 大" }

        // 计算
        if (maxSelectable != null) {
            mGlobalSpec.maxSelectable =
                maxSelectable - (alreadyImageCount + alreadyVideoCount + alreadyAudioCount)
        }
        if (maxImageSelectable != null) {
            mGlobalSpec.maxImageSelectable = maxImageSelectable - alreadyImageCount
        } else {
            mGlobalSpec.maxImageSelectable = null
        }
        if (maxVideoSelectable != null) {
            mGlobalSpec.maxVideoSelectable = maxVideoSelectable - alreadyVideoCount
        } else {
            mGlobalSpec.maxVideoSelectable = null
        }
        if (maxAudioSelectable != null) {
            mGlobalSpec.maxAudioSelectable = maxAudioSelectable - alreadyAudioCount
        } else {
            mGlobalSpec.maxAudioSelectable = null
        }
        return this
    }

    override fun alreadyCount(
        maxSelectable: Int?,
        maxImageSelectable: Int?,
        maxVideoSelectable: Int?,
        maxAudioSelectable: Int?,
        alreadyImageCount: Int,
        alreadyVideoCount: Int,
        alreadyAudioCount: Int
    ): GlobalSetting {
        // 计算
        if (maxSelectable != null) {
            mGlobalSpec.maxSelectable =
                maxSelectable - (alreadyImageCount + alreadyVideoCount + alreadyAudioCount)
        }
        if (maxImageSelectable != null) {
            mGlobalSpec.maxImageSelectable = maxImageSelectable - alreadyImageCount
        }
        if (maxVideoSelectable != null) {
            mGlobalSpec.maxVideoSelectable = maxVideoSelectable - alreadyVideoCount
        }
        if (maxAudioSelectable != null) {
            mGlobalSpec.maxAudioSelectable = maxAudioSelectable - alreadyAudioCount
        }
        return this
    }

    override fun compressImagePath(compressImagePath: String): GlobalSetting {
        mGlobalSpec.compressImagePath = compressImagePath
        return this
    }

    override fun compressVidePath(compressVidePath: String): GlobalSetting {
        mGlobalSpec.compressVidePath = compressVidePath
        return this
    }

    override fun imageEngine(imageEngine: ImageEngine): GlobalSetting {
        mGlobalSpec.imageEngine = imageEngine
        return this
    }

    override fun isCutscenes(isCutscenes: Boolean): GlobalSetting {
        mGlobalSpec.cutscenesEnabled = isCutscenes
        return this
    }

    override fun setRequestedOrientation(@ScreenOrientation requestedOrientation: Int): GlobalSetting {
        mGlobalSpec.orientation = requestedOrientation
        return this
    }

    override fun isImageEdit(isImageEdit: Boolean): GlobalSetting {
        mGlobalSpec.imageEditEnabled = isImageEdit
        return this
    }

    override fun isAddAlbum(byEdit: Boolean): GlobalSetting {
        mGlobalSpec.isAddAlbumByEdit = byEdit
        return this
    }

    override fun setOnImageCompressionListener(listener: OnImageCompressionListener): GlobalSetting {
        mGlobalSpec.onImageCompressionListener = WeakReference(listener).get()
        return this
    }

    override fun setOnLogListener(listener: OnLogListener): GlobalSetting {
        mGlobalSpec.onLogListener = WeakReference(listener).get()
        return this
    }

    override fun videoCompress(videoCompressManager: VideoCompressCoordinator): GlobalSetting {
        mGlobalSpec.videoCompressCoordinator = videoCompressManager
        return this
    }

    override fun forResult(activityResultLauncher: ActivityResultLauncher<Intent>) {
        mGlobalSpec.activityResultLauncher = activityResultLauncher
        openMain(activityResultLauncher)
    }

    /**
     * 调用打开图片、视频预览 - 主要用于配合九宫图
     *
     * @param activity    窗体
     * @param activityResultLauncher 请求器
     * @param list        数据源
     * @param position    当前数据的索引
     * @param isApply 是否有同意操作,支持删除数据源
     */
    override fun openPreviewData(activity: Activity, activityResultLauncher: ActivityResultLauncher<Intent>, list: ArrayList<GridMedia>, position: Int, isApply: Boolean) {
        val intent = Intent(activity, PreviewActivity::class.java)

        // 深度拷贝
        val localMedias = ArrayList<LocalMedia>()
        for (i in list.indices) {
            val localMedia = LocalMedia(list[i])
            localMedias.add(localMedia)
        }

        PreviewSetting(PreviewType.GRID)
            .setLocalMediaArrayList(localMedias)
            .setCurrentPosition(position)
            .isApply(isApply)
            .isOriginal(false)
            .isSelectedCheck(false)
            .isEdit(false)
            .setIntent(intent)
        activityResultLauncher.launch(intent)
        if (GlobalSpec.cutscenesEnabled) {
            activity.overridePendingTransition(R.anim.activity_open_zjh, 0)
        }
    }

    private fun openMain(activityResultLauncher: ActivityResultLauncher<Intent>) {
        val activity = multiMediaSetting.activity ?: return
        // 数量
        var numItems = 0
        // 根据相关配置做相应的初始化
        if (mGlobalSpec.albumSetting != null) {
            numItems++
        }
        if (mGlobalSpec.cameraSetting != null) {
            numItems++
        }
        if (mGlobalSpec.recorderSetting != null) {
            numItems++
        }
        // 如果numItems一个都没，则抛出异常
        check(numItems > 0) { activity.resources.getString(R.string.z_one_of_these_three_albumSetting_camerasSetting_and_recordDerSetting_must_be_set) }
        val intent = Intent(activity, MainActivity::class.java)
        val fragment = multiMediaSetting.fragment
        if (fragment != null) {
            activityResultLauncher.launch(intent)
        } else {
            activityResultLauncher.launch(intent)
            if (mGlobalSpec.cutscenesEnabled) {
                activity.overridePendingTransition(R.anim.activity_open_zjh, 0)
            }
        }
    }

    init {
        mGlobalSpec.setMimeTypeSet(mimeTypes)
    }
}